### PR TITLE
Document native functions for compilation trace

### PR DESCRIPTION
**Pull request by sbaranov**
Fixes https://github.com/flutter/engine/pull/5443#issuecomment-448704175

### CODE REVIEW

File: lib/ui/natives.dart [hunk 1]
```diff
@@ -36,6 +36,10 @@ void _setupHooks() {
   }());
 }
 
+/// Saves Dart compilation trace to a file.
+///
+/// The output is a list of symbols compiled by the Dart JIT at runtime up to this point,
+/// and can be passed later to Dart tools in order to precompile this code offline.
 void saveCompilationTrace(String filePath) {
```
Comment (Hixie) [inline lib/ui/natives.dart]: do we need this if we have `dumpCompilationTrace`?

File: lib/ui/natives.dart [hunk 2]
```diff
@@ -45,6 +49,10 @@ void saveCompilationTrace(String filePath) {
   file.writeAsBytesSync(result);
 }
 
+/// Returns Dart compilation trace as a memory buffer.
+///
+/// The output is a list of symbols compiled by the Dart JIT at runtime up to this point,
+/// and can be passed later to Dart tools in order to precompile this code offline.
 List<int> dumpCompilationTrace() {
```
Comment (Hixie) [inline lib/ui/natives.dart]: the docs here should answer the following questions also:

* what modes (profile, release, debug...) does this do anything useful in
* what happens in the other modes
* what's the actual format of the data?
* what are the tools that you can pass this to?
* can this be depended on long-term?
...and maybe others, i don't know what the questions to ask are. :-)

File: lib/ui/natives.dart (hunk 1 above)
Comment (sbaranov) [reply lib/ui/natives.dart]: Renamed dumpCompilationTrace to saveCompilationTrace. The original saveCompilationTrace was only used by an old internal sample app, and it's OK to break it.

File: lib/ui/natives.dart (hunk 2 above)
Comment (sbaranov) [reply lib/ui/natives.dart]: Done

File: lib/ui/natives.dart [hunk 3]
```diff
@@ -36,15 +36,30 @@ void _setupHooks() {
...
 }
 
-void saveCompilationTrace(String filePath) {
+/// Returns runtime Dart compilation trace as a memory buffer.
+///
+/// The buffer contains a list of symbols compiled by the Dart JIT at runtime up to the point
+/// when this function was called. This list can be saved to a text file and passed to tools
+/// such as `flutter build` or Dart `gen_snapshot` in order to precompile this code offline.
+///
+/// The list has one symbol per line of the following format: `<namespace>,<class>,<symbol>\n`.
```
Comment (Hixie) [inline lib/ui/natives.dart]: one symbol per line implies it's text, but this returns bytes. What's the character encoding? UTF-8?

File: lib/ui/natives.dart [hunk 4]
```diff
@@ -36,15 +36,30 @@ void _setupHooks() {
...
   if (result is Error)
     throw result;
-
-  final File file = new File(filePath);
-  file.writeAsBytesSync(result);
+  return result;
 }
 
+// TODO(sbaranov): This function will go away in subsequent PR, once the framework has caught up.
 List<int> dumpCompilationTrace() {
```
Comment (Hixie) [inline lib/ui/natives.dart]: you can simplify this to `List<int> dumpCompilationTrace() => saveCompilationTrace();`

Comment (Hixie) [reply lib/ui/natives.dart]: might as well mark it `@Deprecated('Use saveCompilationTrace() instead.')` too

**Comment by Hixie** [comment]
We might want to rename this `debugSaveCompilationTrace` if it's only for debug and profile builds. Up to you.

Thanks so much for making the documentation detailed!

LGTM.

File: lib/ui/natives.dart (hunk 3 above)
Comment (sbaranov) [reply lib/ui/natives.dart]: It's text, and I presume it's UTF-8 since it's fetched from Dart VM.

File: lib/ui/natives.dart (hunk 4 above)
Comment (sbaranov) [reply lib/ui/natives.dart]: Done

**Comment by sbaranov** [comment]
This native works fine in dynamic release mode, so renaming it to debugSaveCompilationTrace wouldn't make sense. However, the part in framework that fetches the results of calling this function over service protocol can certainly be renamed debugSaveCompilationTrace.

**Pull request merged**